core/src/main/scala/akka/kafka/internal/CommitObservationLogic.scala

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <https://softwaremill.com>
+ * Copyright (C) 2016 - 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.kafka.internal
+
+import akka.kafka.CommitWhen.{NextOffsetObserved, OffsetFirstObserved}
+import akka.kafka.CommitterSettings
+import akka.kafka.ConsumerMessage.{Committable, CommittableOffset, CommittableOffsetBatch, GroupTopicPartition}
+import akka.stream.stage.GraphStageLogic
+
+/**
+ * Shared commit observation logic between [[GraphStageLogic]] that facilitate offset commits,
+ * such as [[CommitCollectorStage]] and [[CommittingProducerSinkStage]].  It's possible more
+ * logic could be shared between these implementations.
+ */
+private[internal] trait CommitObservationLogic { self: GraphStageLogic =>
+  def settings: CommitterSettings
+
+  /** Batches offsets until a commit is triggered. */
+  protected var offsetBatch: CommittableOffsetBatch = CommittableOffsetBatch.empty
+
+  /** Deferred offsets when `CommitterSetting.when == CommitWhen.NextOffsetObserved` **/
+  private var deferredOffsets: Map[GroupTopicPartition, Committable] = Map.empty
+
+  /**
+   * Update the offset batch when applicable given `CommitWhen` settings. Returns true if the
+   * batch is ready to be committed.
+   */
+  def updateBatch(offset: Committable): Boolean = {
+    val gtp = offset.asInstanceOf[CommittableOffset].partitionOffset.key
+    val updateOffsetBatch = (settings.when, deferredOffsets.get(gtp)) match {
+      case (OffsetFirstObserved, _) =>
+        Some(offset)
+      case (NextOffsetObserved, None) =>
+        deferredOffsets = deferredOffsets + (gtp -> offset)
+        None
+      case (NextOffsetObserved, deferred @ Some(dOffset)) if dOffset != offset =>
+        deferredOffsets = deferredOffsets + (gtp -> offset)
+        deferred
+      case _ => None
+    }
+    for (offset <- updateOffsetBatch) offsetBatch = offsetBatch.updated(offset)
```

Comment:
I'd like to avoid the `Option`s, what about
```scala
    if (settings.when == OffsetFirstObserved) {
      offsetBatch = offsetBatch.updated(offset)
    } else {
      val gtp = offset.asInstanceOf[CommittableOffset].partitionOffset.key
      val dOffset = deferredOffsets.getOrElse(gtp, offset)
      deferredOffsets = deferredOffsets + (gtp -> offset)
      if (dOffset != offset) {
        offsetBatch = offsetBatch.updated(dOffset)
      }
    }
```

Should it be `dOffset < offset`?

Reply:
Ah, I see what you meant now.  Yes, I'll refactor it.

Reply:
LMK what you think now.  I think the pattern matching on the map `get` should be fine.